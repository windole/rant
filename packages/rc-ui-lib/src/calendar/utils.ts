/* eslint-disable no-nested-ternary */
export function compareMonth(date1: Date, date2: Date): number {
  const year1 = date1.getFullYear();
  const year2 = date2.getFullYear();

  if (year1 === year2) {
    const month1 = date1.getMonth();
    const month2 = date2.getMonth();
    return month1 === month2 ? 0 : month1 > month2 ? 1 : -1;
  }

  return year1 > year2 ? 1 : -1;
}

export function compareDay(day1: Date, day2: Date): number {
  const compareMonthResult = compareMonth(day1, day2);

  if (compareMonthResult === 0) {
    const date1 = day1.getDate();
    const date2 = day2.getDate();
    return date1 === date2 ? 0 : date1 > date2 ? 1 : -1;
  }

  return compareMonthResult;
}

export const cloneDate = (date: Date): Date => new Date(date);

export const cloneDates = (dates: Date | Date[]): Date | Date[] =>
  Array.isArray(dates) ? dates.map(cloneDate) : cloneDate(dates);

export function getDayByOffset(date: Date, offset: number): Date {
  const cloned = cloneDate(date);
  cloned.setDate(cloned.getDate() + offset);
  return cloned;
}

export const getPrevDay = (date: Date): Date => getDayByOffset(date, -1);
export const getNextDay = (date: Date): Date => getDayByOffset(date, 1);
export const getToday = (): Date => {
  const today = new Date();
  today.setHours(0, 0, 0, 0);
  return today;
};

export function calcDateNum(date: [Date, Date]): number {
  const day1 = date[0].getTime();
  const day2 = date[1].getTime();
  return (day2 - day1) / (1000 * 60 * 60 * 24) + 1;
}

export const getMonthEndDay = (year: number, month: number): number =>
  32 - new Date(year, month - 1, 32).getDate();

const calendarLang = {
  end: '结束',
  start: '开始',
  title: '日期选择',
  confirm: '确定',
  startEnd: '开始/结束',
  weekdays: ['日', '一', '二', '三', '四', '五', '六'],
  monthTitle: (year: number, month: number) => `${year}年${month}月`,
  rangePrompt: (maxRange: number) => `最多选择 ${maxRange} 天`,
};

export const t = (key: string, ...args): string => {
  if (args.length) {
    return calendarLang[key](...args);
  }
  return calendarLang[key];
};

export const formatMonthTitle = (date: Date): string =>
  t('monthTitle', date.getFullYear(), date.getMonth() + 1);
