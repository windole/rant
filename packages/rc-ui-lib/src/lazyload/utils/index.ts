import { ListenEvent } from '../PropsType';

/* eslint-disable prefer-rest-params */
const inBrowser = typeof window !== 'undefined';

export const DEFAULT_URL =
  'data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7';

// 需要监听的事件
export const DEFAULT_EVENTS: ListenEvent[] = [
  'scroll',
  'wheel',
  'resize',
  'mousewheel',
  'animationend',
  'transitionend',
  'touchmove',
];

export const hasIntersectionObserver =
  inBrowser &&
  'IntersectionObserver' in window &&
  'IntersectionObserverEntry' in window &&
  'intersectionRatio' in window.IntersectionObserverEntry.prototype;

export const modeType = {
  event: 'event',
  observer: 'observer',
};

export const call = (fn): void => fn && fn();

export function setImage(
  element: HTMLMediaElement | HTMLIFrameElement | HTMLImageElement | HTMLElement,
  imagePath: string,
): void {
  if (
    element instanceof HTMLIFrameElement ||
    element instanceof HTMLImageElement ||
    element instanceof HTMLMediaElement
  ) {
    element.src = imagePath;
  } else {
    element.style.backgroundImage = `url('${imagePath}')`;
  }
}

export function loadImage(imagePath: string): Promise<string> {
  return new Promise((resolve, reject) => {
    const img = new Image();
    img.src = imagePath;
    img.onload = () => resolve(imagePath);
    img.onerror = (err) => reject(err);
  });
}

export function addCssClassName(
  element: HTMLImageElement | HTMLElement,
  cssClassName: string,
): void {
  if (!element.className.includes(cssClassName)) {
    element.className += ` ${cssClassName}`;
  }
}

export function throttle(fn, threshold = 250, scope?) {
  let last;
  let deferTimer;
  return () => {
    // @ts-ignore
    const context = scope || this;

    const now = +new Date();
    const args = arguments;
    if (last && now < last + threshold) {
      // hold on to it
      clearTimeout(deferTimer);
      deferTimer = setTimeout(() => {
        last = now;
        fn.apply(context, args);
      }, threshold);
    } else {
      last = now;
      fn.apply(context, args);
    }
  };
}

export function debounce(func, wait, immediate?) {
  let timeout;
  let args;
  let context;
  let timestamp;
  let result;

  const later = function later() {
    const last = +new Date() - timestamp;

    if (last < wait && last >= 0) {
      timeout = setTimeout(later, wait - last);
    } else {
      timeout = null;
      if (!immediate) {
        result = func.apply(context, args);
        if (!timeout) {
          context = null;
          args = null;
        }
      }
    }
  };

  return function debounced() {
    // @ts-ignore
    // eslint-disable-next-line @typescript-eslint/no-this-alias
    context = this;
    args = arguments;
    timestamp = +new Date();

    const callNow = immediate && !timeout;
    if (!timeout) {
      timeout = setTimeout(later, wait);
    }

    if (callNow) {
      result = func.apply(context, args);
      context = null;
      args = null;
    }

    return result;
  };
}
