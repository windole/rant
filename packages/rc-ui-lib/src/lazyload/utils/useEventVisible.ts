import { useEffect, useRef, useState } from 'react';
import { debounce as debounceFunc, DEFAULT_EVENTS, throttle as throttleFunc } from '.';
import { EventOptions } from '../PropsType';

// try to handle passive events
let passiveEventSupported = false;
try {
  const opts = Object.defineProperty({}, 'passive', {
    get() {
      passiveEventSupported = true;
    },
  });
  window.addEventListener('test', null, opts);
  // eslint-disable-next-line no-empty
} catch (e) {}

const passiveEvent = passiveEventSupported ? { capture: false, passive: true } : false;

const useEventVisible = (
  eventOptions: EventOptions,
): [React.MutableRefObject<HTMLDivElement | HTMLImageElement>, boolean] => {
  const {
    listenEvents = DEFAULT_EVENTS,
    offset,
    scrollContainer,
    debounce,
    throttle,
  } = eventOptions;
  // 是否可见
  const [visible, setVisible] = useState(false);
  // // 包裹容器
  const containerRef = useRef<HTMLDivElement | HTMLImageElement>(null);

  useEffect(() => {
    // 检查函数
    let checkVisible = () => {
      // 如果容器不存在则不计算
      if (!containerRef.current) return;
      // 获取当前组件位置
      const { top, left, width, height } = containerRef.current?.getBoundingClientRect();
      // 获取滚动容器位置
      const {
        top: parentTop,
        left: parentLeft,
        width: parentWidth,
        height: parentHeight,
      } = scrollContainer.getBoundingClientRect();
      // 计算屏幕高度和宽度
      const windowInnerHeight = window.innerHeight || document.documentElement.clientHeight;
      const windowInnerWidth = window.innerWidth || document.documentElement.clientWidth;
      // 计算滚动容器和屏幕可视区的相交区域
      const intersectionTop = Math.min(parentTop, 0);
      const intersectionLeft = Math.min(parentLeft, 0);

      const intersectionHeight =
        Math.min(windowInnerHeight, parentTop + parentHeight) - intersectionTop;
      const intersectionWidth =
        Math.min(windowInnerWidth, parentLeft + parentWidth) - intersectionLeft;
      // 计算组件距离可视区的高度
      const offsetTop = top - intersectionTop;
      const offsetLeft = left - intersectionLeft;
      // 偏移量计算[top, left]
      const offsets = Array.isArray(offset) ? offset : [offset, offset];
      // 通过上述距离判断组件是否在可见区域
      const isVisible =
        offsetTop - offsets[0] <= intersectionHeight &&
        offsetTop + height + offsets[0] >= 0 &&
        offsetLeft - offsets[1] <= intersectionWidth &&
        offsetLeft + width + offsets[1] >= 0;

      setVisible(isVisible);
      // 如果可见取消当前组件的所有的监听
      if (isVisible) {
        listenEvents.forEach((event) => {
          scrollContainer?.removeEventListener(event, checkVisible);
        });
      }
    };

    // 使用防抖节流增强checkVisible函数，这里优先防抖，其次节流
    if (debounce) {
      // 防抖函数
      checkVisible = debounceFunc(checkVisible, debounce);
    } else if (throttle) {
      // 节流函数
      checkVisible = throttleFunc(checkVisible, throttle);
    }
    // 批量监听事件
    listenEvents.forEach((event) => {
      scrollContainer?.addEventListener(event, checkVisible, passiveEvent);
    });

    // 自动执行一次
    checkVisible();

    return () => {
      listenEvents.forEach((event) => {
        scrollContainer?.removeEventListener(event, checkVisible, passiveEvent);
      });
    };
  }, [scrollContainer]);
  return [containerRef, visible];
};

export default useEventVisible;
