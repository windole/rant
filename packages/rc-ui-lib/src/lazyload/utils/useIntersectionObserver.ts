import { useCallback, useEffect, useState } from 'react';

const useIntersectionObserver = (
  observerOptions: IntersectionObserverInit,
  handleIntersection: IntersectionObserverCallback,
): ((node: Element) => void) => {
  const [element, setElement] = useState<Element>(null);

  const handleElement = useCallback((node: Element) => {
    setElement(node);
  }, []);

  useEffect(() => {
    if (!element) return;
    const observer = new IntersectionObserver(handleIntersection, observerOptions);
    observer.observe(element);
    // eslint-disable-next-line consistent-return
    return () => observer && observer.disconnect && observer.disconnect();
  }, [handleIntersection, observerOptions, element]);
  return handleElement;
};

export default useIntersectionObserver;
