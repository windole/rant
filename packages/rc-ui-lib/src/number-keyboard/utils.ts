import { KeyConfig, NumberKeyboardProps } from './PropsType';

const genBasicKeys = (props: NumberKeyboardProps) => {
  const keys: KeyConfig[] = Array(9)
    .fill('')
    .map((_, i) => ({ text: i + 1 }));

  if (props.randomKeyOrder) {
    keys.sort(() => (Math.random() > 0.5 ? 1 : -1));
  }

  return keys;
};

export const genDefaultKeys = (props: NumberKeyboardProps): KeyConfig[] => [
  ...genBasicKeys(props),
  { text: props.extraKey as string, type: 'extra' },
  { text: 0 },
  {
    text: props.showDeleteKey ? props.deleteButtonText : '',
    type: props.showDeleteKey ? 'delete' : '',
  },
];

export const genCustomKeys = (props: NumberKeyboardProps) => {
  const keys = genBasicKeys(props);
  const { extraKey } = props;
  const extraKeys = Array.isArray(extraKey) ? extraKey : [extraKey];

  if (extraKeys.length === 1) {
    keys.push({ text: 0, wider: true }, { text: extraKeys[0], type: 'extra' });
  } else if (extraKeys.length === 2) {
    keys.push(
      { text: extraKeys[0], type: 'extra' },
      { text: 0 },
      { text: extraKeys[1], type: 'extra' },
    );
  }

  return keys;
};
