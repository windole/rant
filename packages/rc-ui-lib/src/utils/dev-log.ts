export function devWarning(component: string, message: string): void {
  if (process.env.NODE_ENV !== 'production') {
    console.warn(`[${component}] ${message}`);
  }
}

export function devError(component: string, message: string) {
  if (process.env.NODE_ENV !== 'production') {
    console.error(`[${component}] ${message}`);
  }
}
