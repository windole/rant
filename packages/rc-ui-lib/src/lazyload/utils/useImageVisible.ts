import { useCallback, useState } from 'react';
import { addCssClassName, loadImage, setImage } from '.';
import { IntersectionObserverEntryType, ObserverOptions } from '../PropsType';
import useIntersectionObserver from './useIntersectionObserver';

const defaultObserverOptions = {
  threshold: [0, 1],
};

const useImageVisible = (
  observerOptions: ObserverOptions = defaultObserverOptions,
  errorImage: string,
): [(node: Element) => void, boolean] => {
  const [isLoaded, setIsLoaded] = useState(false);

  const handleIntersection = useCallback(
    (entries: IntersectionObserverEntryType[]) => {
      entries
        .filter((entry) => entry.isIntersecting)
        .forEach((entry) => {
          const target = entry.target as HTMLImageElement | HTMLElement;
          const url = target.dataset.src || '';
          loadImage(url)
            .catch(() => {
              setImage(target, errorImage);
              addCssClassName(target, 'rc-lazyload__error');
              target.setAttribute('lazy', 'error');
            })
            .then((imagePath: string) => {
              setImage(target, imagePath);
              setIsLoaded(true);
              addCssClassName(target, 'rc-lazyload__loaded');
              target.setAttribute('lazy', 'loaded');
              target.removeAttribute('data-src');
            });
        });
    },
    [setIsLoaded, isLoaded],
  );
  const ref = useIntersectionObserver(observerOptions, handleIntersection);
  return [ref, isLoaded];
};

export default useImageVisible;
