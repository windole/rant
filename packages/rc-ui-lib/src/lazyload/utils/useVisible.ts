import { useCallback, useEffect, useState } from 'react';
import { addCssClassName } from '.';
import useIntersectionObserver from './useIntersectionObserver';

const defaultObserverOptions = {
  threshold: [0, 1],
};

const useVisible = (
  initialVisible: boolean,
  observerOptions = defaultObserverOptions,
): [(node: Element) => void, boolean] => {
  const [visible, setVisible] = useState(initialVisible);

  const forceVisible = () => {
    setVisible(true);
  };

  useEffect(() => {
    if (initialVisible) {
      forceVisible();
    }
  }, [initialVisible]);

  const handleIntersection = useCallback(
    (entries) => {
      entries
        .filter((entry) => entry.isIntersecting)
        .forEach((entry) => {
          const target = entry.target as HTMLElement;
          addCssClassName(target, 'rc-lazyload__loaded');
          setVisible(true);
        });
    },
    [setVisible, visible],
  );
  const ref = useIntersectionObserver(observerOptions, handleIntersection);
  return [ref, visible];
};

export default useVisible;
