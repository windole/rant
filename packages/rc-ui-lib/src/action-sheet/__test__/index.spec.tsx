import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { ActionSheet } from '..';
import { sleep } from '../../../tests/utils';

describe('Popup', () => {
  let wrapper;

  afterEach(() => {
    wrapper.unmount();
  });

  test('should emit select event after clicking option', async () => {
    // const wrapper = mount(ActionSheet, {
    //     props: {
    //         show: true,
    //         actions: [{ name: 'Option' }],
    //     },
    // });
    const actions = [{ name: 'Option' }];
    const onSelect = jest.fn();
    wrapper = mount(<ActionSheet visible actions={actions} onSelect={onSelect} />);
    wrapper.find('.rc-action-sheet__item').simulate('click');
    await sleep(0);
    expect(onSelect).toHaveBeenCalledTimes(1);
    expect(onSelect).toHaveBeenCalledWith(
      {
        name: 'Option',
      },
      0,
    );
  });
  test('should call callback function after clicking option', () => {
    const callback = jest.fn();
    const actions = [{ name: 'Option', callback }];
    wrapper = mount(<ActionSheet visible actions={actions} />);
    wrapper.find('.rc-action-sheet__item').simulate('click');
    expect(callback).toHaveBeenCalledTimes(1);
  });

  test('should not emit select event after clicking loading option', async () => {
    const onSelect = jest.fn();
    const actions = [{ name: 'Option', loading: true }];
    wrapper = mount(<ActionSheet visible actions={actions} onSelect={onSelect} />);
    wrapper.find('.rc-action-sheet__item').simulate('click');
    await sleep(0);
    expect(onSelect).not.toHaveBeenCalled();
  });

  test('should not emit select event after clicking disabled option', async () => {
    const onSelect = jest.fn();
    const actions = [{ name: 'Option', disabled: true }];
    wrapper = mount(<ActionSheet visible actions={actions} onSelect={onSelect} />);
    wrapper.find('.rc-action-sheet__item').simulate('click');
    await sleep(0);
    expect(onSelect).not.toHaveBeenCalled();
  });

  test('should emit cancel event after clicking cancel button', () => {
    const onCancel = jest.fn();
    const actions = [{ name: 'Option' }];
    wrapper = mount(
      <ActionSheet visible actions={actions} cancelText="Cancel" onCancel={onCancel} />,
    );
    wrapper.find('.rc-action-sheet__cancel').simulate('click');
    expect(onCancel).toHaveBeenCalledTimes(1);
  });

  test('should render subname correctly', () => {
    const actions = [{ name: 'Option', subname: 'Subname' }];
    wrapper = mount(<ActionSheet visible actions={actions} cancelText="Cancel" />);
    expect(wrapper.find('.rc-action-sheet__subname').text()).toEqual('Subname');
    expect(wrapper.find('.rc-action-sheet__item').html()).toMatchSnapshot();
  });

  test('should render default slot correctly', () => {
    wrapper = mount(
      <ActionSheet visible title="Title">
        Default
      </ActionSheet>,
    );
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('should have "van-popup--round" class when setting the round prop', () => {
    wrapper = mount(<ActionSheet visible round />);
    expect(wrapper.find('.rc-popup--round').exists()).toBeTruthy();
  });

  test('should change option color when using the color prop', () => {
    const actions = [{ name: 'Option', color: 'red' }];
    wrapper = mount(<ActionSheet visible actions={actions} />);
    const item = wrapper.find('.rc-action-sheet__item');
    expect(item.getDOMNode().style.color).toEqual('red');
  });

  test('should hide close icon when the closeable prop is false', async () => {
    wrapper = mount(<ActionSheet visible title="Title" />);
    expect(wrapper.find('.rc-action-sheet__close').exists()).toBeTruthy();

    await wrapper.setProps({ closeable: false });
    expect(wrapper.find('.rc-action-sheet__close').exists()).toBeFalsy();
  });

  test('should allow to custom close icon with closeIcon prop', () => {
    wrapper = mount(<ActionSheet visible title="Title" closeIcon="success" />);
    expect(toJson(wrapper.find('.rc-action-sheet__close'))).toMatchSnapshot();
  });

  test('should render description correctly', () => {
    wrapper = mount(<ActionSheet visible description="This is a description" />);
    expect(wrapper.find('.rc-action-sheet__description').text()).toEqual('This is a description');
    expect(toJson(wrapper.find('.rc-action-sheet__description'))).toMatchSnapshot();
  });

  // test('should close after clicking option if close-on-click-action prop is true', () => {
  //     const actions = [{ name: 'Option' }];
  //     wrapper = mount(<ActionSheet visible closeOnClickAction actions={actions} />);
  //     wrapper.find('.rc-action-sheet__item').simulate('click');
  //     console.log("============", wrapper.find(".rc-action-sheet").at(1).html());
  //     expect(wrapper.find(".rc-action-sheet").at(1).getDOMNode().style.display).toEqual("none");
  // });

  test('should emit click-overlay event and closed after clicking the overlay', () => {
    const onClickOverlay = jest.fn();
    const actions = [{ name: 'Option' }];
    wrapper = mount(<ActionSheet visible onClickOverlay={onClickOverlay} actions={actions} />);

    wrapper.find('.rc-overlay').simulate('click');
    // expect(wrapper.emitted('update:show')![0]).toEqual([false]);
    expect(onClickOverlay).toHaveBeenCalledTimes(1);
  });

  test('should allow to control safe-area with safe-area-inset-bottom prop', async () => {
    wrapper = mount(<ActionSheet visible />);
    expect(wrapper.find('.rc-action-sheet').at(1).hasClass('rc-safe-area-bottom')).toBeTruthy();

    await wrapper.setProps({
      safeAreaInsetBottom: false,
    });
    expect(wrapper.find('.rc-action-sheet').at(1).hasClass('rc-safe-area-bottom')).toBeFalsy();
  });
});
