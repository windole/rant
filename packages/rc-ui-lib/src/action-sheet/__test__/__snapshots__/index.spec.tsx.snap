// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popup should allow to custom close icon with closeIcon prop 1`] = `
Array [
  <Icon
    className="rc-action-sheet__close"
    name="success"
    onClick={[Function]}
    tag="i"
  >
    <Badge
      className="rc-action-sheet__close van-icon van-icon-success"
      onClick={[Function]}
      showZero={true}
      style={
        Object {
          "color": undefined,
          "fontSize": undefined,
        }
      }
      tag="i"
    >
      <i
        className="rc-badge__wrapper rc-action-sheet__close van-icon van-icon-success"
        onClick={[Function]}
        style={
          Object {
            "color": undefined,
            "fontSize": undefined,
          }
        }
      />
    </Badge>
  </Icon>,
  <Badge
    className="rc-action-sheet__close van-icon van-icon-success"
    onClick={[Function]}
    showZero={true}
    style={
      Object {
        "color": undefined,
        "fontSize": undefined,
      }
    }
    tag="i"
  >
    <i
      className="rc-badge__wrapper rc-action-sheet__close van-icon van-icon-success"
      onClick={[Function]}
      style={
        Object {
          "color": undefined,
          "fontSize": undefined,
        }
      }
    />
  </Badge>,
  <i
    className="rc-badge__wrapper rc-action-sheet__close van-icon van-icon-success"
    onClick={[Function]}
    style={
      Object {
        "color": undefined,
        "fontSize": undefined,
      }
    }
  />,
]
`;

exports[`Popup should render default slot correctly 1`] = `"<div style=\\"z-index: 2000; animation-duration: 300ms;\\" class=\\"rc-overlay\\"></div><div style=\\"z-index: 2000; transition-duration: 300ms;\\" class=\\"rc-popup rc-popup--round rc-popup--bottom rc-safe-area-bottom rc-action-sheet\\"><div class=\\"rc-action-sheet__header\\">Title<i class=\\"rc-badge__wrapper rc-action-sheet__close van-icon van-icon-cross\\"></i></div><div class=\\"rc-action-sheet__content\\">Default</div></div>"`;

exports[`Popup should render description correctly 1`] = `
<div
  className="rc-action-sheet__description"
>
  This is a description
</div>
`;

exports[`Popup should render subname correctly 1`] = `"<button type=\\"button\\" class=\\"rc-action-sheet__item\\"><span class=\\"rc-action-sheet__name\\">Option</span><div class=\\"rc-action-sheet__subname\\">Subname</div></button>"`;
